File: AgeCalculator/AgeCalculator/AgeResult.swift
import Foundation

struct AgeResult {
    let years: Int
    let months: Int
    let days: Int
    let totalMonths: Int
    let totalDays: Int

    static func calculate(from dob: Date, to today: Date = Date()) -> AgeResult? {
        guard dob <= today else { return nil }

        let calendar = Calendar.current
        let components = calendar.dateComponents([.year, .month, .day], from: dob, to: today)

        let years = components.year ?? 0
        let months = components.month ?? 0
        let days = components.day ?? 0

        let totalMonthsComponents = calendar.dateComponents([.month], from: dob, to: today)
        let totalMonths = totalMonthsComponents.month ?? 0

        let totalDaysComponents = calendar.dateComponents([.day], from: dob, to: today)
        let totalDays = totalDaysComponents.day ?? 0

        return AgeResult(
            years: years,
            months: months,
            days: days,
            totalMonths: totalMonths,
            totalDays: totalDays
        )
    }
}

File: AgeCalculatorFlutter/lib/age_result.dart
class AgeResult {
  final int years;
  final int months;
  final int days;
  final int totalMonths;
  final int totalDays;

  const AgeResult({
    required this.years,
    required this.months,
    required this.days,
    required this.totalMonths,
    required this.totalDays,
  });

  static AgeResult? calculate(DateTime dob) {
    final today = DateTime.now();
    final todayDate = DateTime(today.year, today.month, today.day);
    final dobDate = DateTime(dob.year, dob.month, dob.day);

    if (dobDate.isAfter(todayDate)) return null;

    int years = todayDate.year - dobDate.year;
    int months = todayDate.month - dobDate.month;
    int days = todayDate.day - dobDate.day;

    if (days < 0) {
      months -= 1;
      final prevMonth = DateTime(todayDate.year, todayDate.month, 0);
      days += prevMonth.day;
    }
    if (months < 0) {
      years -= 1;
      months += 12;
    }

    int totalMonths = (todayDate.year - dobDate.year) * 12 +
        (todayDate.month - dobDate.month);
    if (todayDate.day < dobDate.day) totalMonths -= 1;

    final totalDays = todayDate.difference(dobDate).inDays;

    return AgeResult(
      years: years,
      months: months,
      days: days,
      totalMonths: totalMonths,
      totalDays: totalDays,
    );
  }
}
